### PR TITLE
[Caffe2 CPU tests] Update CMakeLists.txt

### PR DESCRIPTION
**Pull request by kunalb**
I was trying to build PyTorch with USE_GLOG=ON (so we could get better timestamps around the nccl logging) and ran into this error

```
[1/7] Linking CXX executable bin/verify_api_visibility
FAILED: bin/verify_api_visibility
: && /opt/rh/gcc-toolset-11/root/usr/bin/c++ -D_GLIBCXX_USE_CXX11_ABI=1 -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DUSE_KINETO -DLIBKINETO_NOROCTRACER -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=range-loop-construct -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-stringop-overflow -Wsuggest-override -Wno-psabi -Wno-error=pedantic -Wno-error=old-style-cast -Wno-missing-braces -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow -DHAVE_AVX512_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O2 -g -DNDEBUG -rdynamic     -Wl,--no-as-needed caffe2/CMakeFiles/verify_api_visibility.dir/__/aten/src/ATen/test/verify_api_visibility.cpp.o -o bin/verify_api_visibility -L/lib/intel64   -L/lib/intel64_win   -L/lib/win-x64 -Wl,-rpath,/lib/intel64:/lib/intel64_win:/lib/win-x64:/usr/local/cuda/lib64:/root/conda/lib:/mnt/code/pytorch/build/lib:  lib/libgtest_main.a  -Wl,--no-as-needed,"/mnt/code/pytorch/build/lib/libtorch.so" -Wl,--as-needed  -Wl,--no-as-needed,"/mnt/code/pytorch/build/lib/libtorch_cpu.so" -Wl,--as-needed  lib/libprotobuf.a  /root/conda/lib/libmkl_intel_lp64.so  /root/conda/lib/libmkl_gnu_thread.so  /root/conda/lib/libmkl_core.so  -fopenmp  /usr/lib64/libpthread.so  -lm  /usr/lib64/libdl.so  -Wl,--no-as-needed,"/mnt/code/pytorch/build/lib/libtorch_cuda.so" -Wl,--as-needed  lib/libc10_cuda.so  lib/libc10.so  /root/conda/lib/libglog.so.0.4.0  /root/conda/lib/libgflags.so.2.2.2  -lpthread  /usr/local/cuda/lib64/libcudart.so  /usr/local/cuda/lib64/libnvToolsExt.so  lib/libgtest.a  -pthread && /root/conda/bin/cmake -E __run_co_compile --lwyu="ldd;-u;-r" --source=bin/verify_api_visibility && :
/opt/rh/gcc-toolset-11/root/usr/bin/ld: /mnt/code/pytorch/build/lib/libtorch.so: undefined reference to symbol '_ZTVN10__cxxabiv117__class_type_infoE@@CXXABI_1.3'
/opt/rh/gcc-toolset-11/root/usr/bin/ld: /usr/lib64/libstdc++.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```

Adding stdc++ explicitly to the list of libraries to link seems to fix the build, and I was able to get a working build of PyTorch.

### CODE REVIEW

**Review by zdevito — APPROVED** [review]

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -1781,7 +1781,7 @@ if(BUILD_TEST)
   foreach(test_src ${Caffe2_CPU_TEST_SRCS})
     get_filename_component(test_name ${test_src} NAME_WE)
     add_executable(${test_name} "${test_src}")
-    target_link_libraries(${test_name} torch_library gtest_main)
+    target_link_libraries(${test_name} torch_library gtest_main stdc++)
```
Comment (malfet) [inline caffe2/CMakeLists.txt:1784]: At the very least you need to do this, as stdc++ is not available on Windows
```suggestion
if(not MSVC)
    target_link_libraries(${test_name} torch_library gtest_main stdc++)
endif()
```

**Comment by malfet** [comment]
Added a guard against not-windows (stdc++ is a very linux library), but this change feels very suspicious to me as dependency to C++ runtime should be added by automatically for C++ projects, and libtorch is one of them

File: caffe2/CMakeLists.txt (hunk 1 above)
Comment (cyyever) [reply caffe2/CMakeLists.txt:1785]: No, we should use 
```
 target_link_libraries(${test_name} torch_library gtest_main  )
if(NOT MSVC)
    target_link_libraries(${test_name}  stdc++)
endif()
```

**Comment by cyyever** [comment]
I suspect that gcc-11 was used in CentOS7 so it has the path '/opt/rh/gcc-toolset-11/', and the system's default libstdc++ may have older ABI. May be the proper libstdc++ with /opt/rh/gcc-toolset-11/ should be installed and CMake should be able to find it. (By using some kind of environment variable?)

**Comment by kunalb** [comment]
Hm, I was using a full toolchain installed into centos 8 (gcc-toolset-11-11.1-1.el8.x86_64), then run with `scl enable gcc-toolset-11` to have everything pointed to the right location.

I'll double check

**Comment by malfet** [comment]
@kunalb I don't think there is a big harm in adding libstdc++ as a dependency, so feel free to land it (windows builds are passing now and runtime failure is unrelated), but again this dependency feels a bit superfluous though not wrong

**Comment by kunalb** [comment]
Thanks, I'll go ahead and land it in that case and then keep looking on the side.

@pytorchmergebot merge

**Pull request closed without merging**